File: packages/sdk-python/examples/pkce_auth_code.py
"""
Kang Open Banking — minimal PKCE Authorization Code sample (Python 3.10+).

Run:
    export KOB_BASE=https://sandbox-api.kangopenbanking.com/v1
    export KOB_CLIENT_ID=<your_sandbox_client_id>
    export KOB_REDIRECT_URI=http://127.0.0.1:8765/callback
    export KOB_SCOPE="openid accounts"
    python packages/sdk-python/examples/pkce_auth_code.py

Steps:
    1. Generate a 64-byte code_verifier and S256 code_challenge.
    2. Print the /oauth/authorize URL.  Open it in any browser, sign in,
       consent.
    3. Capture the redirect on 127.0.0.1:8765, validate state, exchange the
       code at /oauth/token with grant_type=authorization_code (public client,
       no client_secret).
    4. Call /health as a sanity probe with the bearer.

RFC 7636 §4 (PKCE).  Stdlib only — uses http.server, urllib, secrets, hashlib.
"""
from __future__ import annotations

import base64
import hashlib
import http.server
import os
import secrets
import sys
import threading
import urllib.parse
import urllib.request


BASE = os.environ.get("KOB_BASE", "https://sandbox-api.kangopenbanking.com/v1")
CLIENT_ID = os.environ.get("KOB_CLIENT_ID")
REDIRECT_URI = os.environ.get("KOB_REDIRECT_URI", "http://127.0.0.1:8765/callback")
SCOPE = os.environ.get("KOB_SCOPE", "openid accounts")

if not CLIENT_ID:
    sys.exit("KOB_CLIENT_ID is required.")


def b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


code_verifier = b64url(secrets.token_bytes(64))
code_challenge = b64url(hashlib.sha256(code_verifier.encode()).digest())
state = b64url(secrets.token_bytes(16))

params = {
    "response_type": "code",
    "client_id": CLIENT_ID,
    "redirect_uri": REDIRECT_URI,
    "scope": SCOPE,
    "state": state,
    "code_challenge": code_challenge,
    "code_challenge_method": "S256",
}
print("\nOpen this URL in your browser, sign in, and consent:\n")
print(f"{BASE}/oauth/authorize?{urllib.parse.urlencode(params)}\n")


captured: dict[str, str] = {}


class Handler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):  # noqa: N802
        parsed = urllib.parse.urlparse(self.path)
        if parsed.path != urllib.parse.urlparse(REDIRECT_URI).path:
            self.send_response(404)
            self.end_headers()
            return
        captured.update(dict(urllib.parse.parse_qsl(parsed.query)))
        self.send_response(200)
        self.send_header("content-type", "text/html")
        self.end_headers()
        self.wfile.write(
            b"<h2>Kang Open Banking - code captured. You can close this tab.</h2>"
        )
        threading.Thread(target=self.server.shutdown, daemon=True).start()

    def log_message(self, *_):  # silence
        pass


parsed = urllib.parse.urlparse(REDIRECT_URI)
server = http.server.HTTPServer((parsed.hostname or "127.0.0.1", parsed.port or 8765), Handler)
server.serve_forever()

if captured.get("state") != state:
    sys.exit("state mismatch - possible CSRF")
code = captured.get("code")
if not code:
    sys.exit("no code returned")

print("Received authorization code, exchanging for token...")
body = urllib.parse.urlencode(
    {
        "grant_type": "authorization_code",
        "code": code,
        "client_id": CLIENT_ID,
        "redirect_uri": REDIRECT_URI,
        "code_verifier": code_verifier,
    }
).encode()

req = urllib.request.Request(
    f"{BASE}/oauth/token",
    data=body,
    headers={"content-type": "application/x-www-form-urlencoded"},
)
import json
with urllib.request.urlopen(req, timeout=15) as resp:
    token = json.loads(resp.read())
print(f"Got access token (expires in {token.get('expires_in')}s).")

probe = urllib.request.Request(
    f"{BASE}/health",
    headers={"authorization": f"Bearer {token['access_token']}"},
)
with urllib.request.urlopen(probe, timeout=15) as resp:
    print("Secured /health call returned", resp.status)

File: packages/sdk-node/examples/pkce-auth-code.ts
/**
 * Kang Open Banking — minimal PKCE Authorization Code sample (Node 18+).
 *
 * Run:
 *   export KOB_BASE=https://sandbox-api.kangopenbanking.com/v1
 *   export KOB_CLIENT_ID=<your_sandbox_client_id>
 *   export KOB_REDIRECT_URI=http://127.0.0.1:8765/callback
 *   export KOB_SCOPE="openid accounts payments"
 *   bun packages/sdk-node/examples/pkce-auth-code.ts
 *
 * The script:
 *   1. Generates a 64-byte code_verifier and S256 code_challenge.
 *   2. Prints the /oauth/authorize URL — open it in any browser, sign in,
 *      consent.
 *   3. Listens on http://127.0.0.1:8765/callback for the redirect, captures
 *      ?code=&state= and validates state.
 *   4. Exchanges the code at /oauth/token with grant_type=authorization_code
 *      and the verifier (no client_secret — public client).
 *   5. Calls /health as a sanity probe with the returned bearer.
 *
 * RFC 7636 §4 (PKCE). No third-party deps; uses Node's built-in http + crypto.
 */
import { createServer } from 'node:http';
import { randomBytes, createHash } from 'node:crypto';
import { URL } from 'node:url';

const BASE = process.env.KOB_BASE ?? 'https://sandbox-api.kangopenbanking.com/v1';
const CLIENT_ID = process.env.KOB_CLIENT_ID;
const REDIRECT_URI = process.env.KOB_REDIRECT_URI ?? 'http://127.0.0.1:8765/callback';
const SCOPE = process.env.KOB_SCOPE ?? 'openid accounts';

if (!CLIENT_ID) {
  console.error('KOB_CLIENT_ID is required.');
  process.exit(1);
}

const b64url = (buf: Buffer) =>
  buf.toString('base64').replace(/=+$/, '').replace(/\+/g, '-').replace(/\//g, '_');

const codeVerifier = b64url(randomBytes(64));
const codeChallenge = b64url(createHash('sha256').update(codeVerifier).digest());
const state = b64url(randomBytes(16));

const authorizeUrl = new URL(`${BASE}/oauth/authorize`);
authorizeUrl.searchParams.set('response_type', 'code');
authorizeUrl.searchParams.set('client_id', CLIENT_ID);
authorizeUrl.searchParams.set('redirect_uri', REDIRECT_URI);
authorizeUrl.searchParams.set('scope', SCOPE);
authorizeUrl.searchParams.set('state', state);
authorizeUrl.searchParams.set('code_challenge', codeChallenge);
authorizeUrl.searchParams.set('code_challenge_method', 'S256');

console.log('\nOpen this URL in your browser, sign in, and consent:\n');
console.log(authorizeUrl.toString(), '\n');

const code: string = await new Promise((resolve, reject) => {
  const server = createServer((req, res) => {
    if (!req.url) return;
    const url = new URL(req.url, REDIRECT_URI);
    if (url.pathname !== new URL(REDIRECT_URI).pathname) {
      res.writeHead(404).end();
      return;
    }
    const returnedState = url.searchParams.get('state');
    const returnedCode = url.searchParams.get('code');
    res.writeHead(200, { 'content-type': 'text/html' });
    res.end('<h2>Kang Open Banking — code captured. You can close this tab.</h2>');
    server.close();
    if (returnedState !== state) return reject(new Error('state mismatch — possible CSRF'));
    if (!returnedCode) return reject(new Error('no code returned'));
    resolve(returnedCode);
  });
  const { port, hostname } = new URL(REDIRECT_URI);
  server.listen(Number(port), hostname);
});

console.log('Received authorization code, exchanging for token...');

const tokenRes = await fetch(`${BASE}/oauth/token`, {
  method: 'POST',
  headers: { 'content-type': 'application/x-www-form-urlencoded' },
  body: new URLSearchParams({
    grant_type: 'authorization_code',
    code,
    client_id: CLIENT_ID,
    redirect_uri: REDIRECT_URI,
    code_verifier: codeVerifier,
  }),
});

if (!tokenRes.ok) {
  console.error('token exchange failed:', tokenRes.status, await tokenRes.text());
  process.exit(1);
}
const token = await tokenRes.json() as { access_token: string; token_type: string; expires_in: number };
console.log('Got access token (expires in', token.expires_in, 's).');

const probe = await fetch(`${BASE}/health`, {
  headers: { authorization: `Bearer ${token.access_token}` },
});
console.log('Secured /health call returned', probe.status);
process.exit(probe.ok ? 0 : 1);
